lib/resty/rediscluster.lua

```diff
@@ -714,63 +781,80 @@ function _M.commit_pipeline(self)
     slots = nil -- luacheck: ignore
 
     for k, v in pairs(node_req_map) do
-        local ip = v.ip
-        local port = v.port
         local reqs = v.reqs
-        local slave = v.slave
-        local redis_client = redis:new()
-        redis_client:set_timeouts(config.connect_timeout or DEFAULT_CONNECTION_TIMEOUT,
-                                  config.send_timeout or DEFAULT_SEND_TIMEOUT,
-                                  config.read_timeout or DEFAULT_READ_TIMEOUT)
-        local ok, err = redis_client:connect(ip, port, self.config.connect_opts)
+
+        local ip, port
+        local redis_client
+        local ok, err_t = nil, {}
+        -- try ip first and then fallback to hostname
+        for i = 1, CONSTANT_TWO do
+            local slave, err
+            ip, port, slave, err = pick_node(self, v.serv_list, magicRandomPickupSeed, i%CONSTANT_TWO + 1)
```

Comment:
What is the `i%CONSTANT_TWO + 1` supposed to do?  It is not at all obvious and requires too much thinking.  Instead of a `for` loop, factoring out its body into a separate function and that is then called once or maybe twice, if the first call did not yield a connection, would be way easier to understand.

Reply:
I have re-done the code.